proposals/compile-time-dependency-resolution.md

```diff
@@ -0,0 +1,590 @@
+
+## Composition and chain of evidences
+
+Constraint interface declarations and extension evidences can encode further constraints on their type parameters so that they can be composed nicely:
+
+```kotlin
+package com.data.instances
+
+import com.data.Repository
+import com.domain.User
+
+data class Group<A>(val values: List<A>)
+
+extension class GroupRepository<A>(with val repoA: Repository<A>) : Repository<Group<A>> {
+  override fun loadAll(): Map<Int, Group<A>> {
+    return repoA.loadAll().mapValues { Group(it.value) }
+  }
+
+  override fun loadById(id: Int): Group<A>? {
+    return repoA.loadById(id)?.let { Group(it) }
+  }
+
+  override fun Group<A>.save() {
+    this.items.map { repoA.run { it.save() } }
+  }
+}
+```
+
+The above extension provides evidence of a `Repository<Group<A>>` as long as there is a `Repository<A>` in scope. The Call site would look like:
+
+```kotlin
+fun <A> fetchGroup(with repo: GroupRepository<A>) = loadAll()
+
+fun main() {
+  fetchGroup<User>() // Succeeds! There's evidence of Repository<Group<A>> and Repository<User> provided in scope.
+  fetchGroup<Coin>() // Fails! There's evidence of Repository<Group<A>> but no evidence of `Repository<Coin>` available.
+}
+```
+
+We believe that the encoding proposed above fits nicely with Kotlin's philosophy of extensions and will reduce the boilerplate compared to other languages that also support compile-time dependency resolution.
+
+
+## Language changes
+
+* Add `with` to require evidence of extensions in both function and class/object declarations.
+* Add `extension` to provide instance evidence for a given interface.
+
+Usage of these language changes are demonstrated by the previous and below examples:
+
+#### Class constraint
+
+```kotlin
+extension class GroupRepository<A>(with R: Repository<A>) : Repository<Group<A>> {
+  /* ... */
+}
+```
+
+#### Function constraint
+
+```kotlin
+fun <A> fetch(id: String, with R: Repository<A>): A = loadById(id) // function position using parameter "R"
+```
+
+#### Extension evidence using an Object
+
+```kotlin
+extension object UserRepository : Repository<User> {
+
+  val storedUsers: MutableMap<Int, User> = mutableMapOf() // e.g. users stored in a DB
+
+  override fun loadAll(): Map<Int, User> {
+    return storedUsers
+  }
+
+  override fun loadById(id: Int): User? {
+    return storedUsers[id]
+  }
+
+  override fun User.save() {
+    storedUsers[this.id] = this
+  }
+}
+```
+
+#### Extension evidence using a Class
+
+```kotlin
+extension class UserRepository: Repository<User> {
+
+  val storedUsers: MutableMap<Int, User> = mutableMapOf() // e.g. users stored in a DB
+
+  override fun loadAll(): Map<Int, User> {
+    return storedUsers
+  }
+
+  override fun loadById(id: Int): User? {
+    return storedUsers[id]
+  }
+
+  override fun User.save() {
+    storedUsers[this.id] = this
+  }
+}
+```
+
+## Extension resolution order
```

Comment:
Why is a resolution order necessary? Shouldn't overlapping instances be straight out rejected?

Assuming that extensions are resolved according to the resolution order, are they then resolved using backtracking search? This seems like an important detail in the presence of extension classes with extension constraints. For example:
```kotlin
extension interface Foo<A> {
  companion object {
    extension class MyClassFoo(): Foo<MyClass> {}
  }
}
class MyClass() {
  companion object {
    extension class PriorityFoo(with val unsatisfiable: Foo<Unit>): Foo<MyClass> {}
  }
}

fun example(with myClassFoo: Foo<MyClass) = ...
```

In this example, will a call `example()` result in a type error or will it get resolved to `example(Foo.MyClassFoo())`? The former will happen if there is no backtracking, because the extension class on the companion object of the target type gets priority, but its own constraints are not satisfiable (no instance of `Foo<Unit>`). In the latter case, backtracking will continue the search among instances with lesser priority, and will thus find the instance on the companion object of the extension interface.

Reply:
The resolution order makes it easier to understand - for the developer, and potentially faster for the compiler.

Do you have a more realistic example where backtracking would make sense?

Reply:
I don't actually think backtracking is a good idea since it will make the process of resolving constraints very hard to follow - my comment was meant as a question to clarify whether the KEEP actually has backtracking in mind or not.

I don't understand the point about the resolution order making it easier for the programmer to understand and faster for the compiler. I understand the resolution order as a mechanism for resolving *ambiguity* when more than one extension class matches a given constraint. My question is: why is this ambiguity even allowed to occur? That is, shouldn't overlapping instances be rejected by the compiler? If overlapping instances are rejected, then there is no need for a resolution order because there is at most one extension class that matches.

Reply:
(Replying to https://github.com/Kotlin/KEEP/pull/87#issuecomment-499092718 in this thread)

> @zeitlinger
> Why allow ambiguity? Think if the Equality use case as described in https://quickbirdstudios.com/blog/keep-87-typeclasses-kotlin/  Equality based on the dog interface (based on name) Equality based on the database dog implementation, which implements the dog interface  Seems to make sense to me, but would be rejected according to your proposal.

That is actually an excellent example of why one would want to disallow overlapping instances :).

First of all, and this is not my main point, the premise of the article is a bit misguided, as it assumes that one would actually want to have an `equals()` method which compares specific implementations (`DatabaseDog`) according to one equivalence relation and use a "default" equivalence relation for everything else that happens to implement the `Dog` interface. This is a variant of the common trap of trying to implement `equals()` for open class hierarchies: it is *impossible*, in general, to implement `equals()` in this way because one of the required properties of symmetry or transitivity will fail. For example, we have `DatabaseDog(1, "Klaus") == ApiDog("Klaus")` and `ApiDog("Klaus") == DatabaseDog(2, "Klaus")` but `DatabaseDog(1, "Klaus") != DatabaseDog(2, "Klaus")`, so transitivity is broken (and it cannot be fixed).

Getting back to type classes, we put these problems aside and ignore the actual semantics of what the author is trying to implement. He argues here that it is a good idea to be able to define a "default" instance for the general interface `Dog` and override it with a specific one for the concrete class `DatabaseDog`. But this is really dangerous! Remember that the target type used for looking up an extension class is the type of the object at *compile time*, and not the actual run-time type. This means that in the following example:
```
val dog1 = DatabaseDog(1, "Klaus")
val dog2 = DatabaseDog(2, "Klaus")
val b = dog1.eq(dog2) // evaluates to false
```
the compiler will infer `DatabaseDog` for the two variables `dog1` and `dog2`, and `b` will get the value `false` because the ids of the two objects are not equal.
However, just by adding annotations on the first object, we can completely change the semantics of the program:
```
val dog1: Dog = DatabaseDog(1, "Klaus")
val dog2 = DatabaseDog(2, "Klaus")
val b = dog1.eq(dog2) //evaluates to true
```
This is because now the extension class for `DatabaseDog` no longer matches, and the fallback extension class defined for `Dog` is used instead. None of the actual objects changed, only the type annotation! Reified type parameters aside, this would, to my knowledge, be the only place where the semantics of a Kotlin program would be affected by the static types of the objects. Type inference means that types are rarely mentioned, making it really difficult to determine what implementation is used just by looking at the code. Worse, an apparently safe refactoring could cause an inferred type to be made more general than before, subtly changing the program in critical ways.

Haskell, another language with type classes, will fail at compile time if a constraint can be satisfied by more than one instance of a type class for the same reasons as I have stated above. And Haskell does not even have subtyping, which further increases the possible sources of errors.

I would strongly suggest disallowing overlapping instances. It can always be added back, but once it is the default behavior, it is going to be very hard to remove it again. **If** they are to be allowed, I suggest going in the direction of Haskell and require the programmer to judiciously add explicit annotations (`{-# OVERLAPPING #-}`, `-XIncoherentInstances`, etc.) if the design really cannot be changed to not use overlapping instances, e.g. because of legacy code.

Reply:
> Reified type parameters aside, this would, to my knowledge, be the only place where the semantics of a Kotlin program would be affected by the static types of the objects.

Not true. This is auctually quite common, and it is called function overloading. Overloaded functions are statically resolved, i.e. via the static compile-time type of their arguments. Example:

```
fun whatDog(dog: Dog) = print("dog")
fun whatDog(dog: DatabaseDog) = print("database dog")

fun test() {
    val dog = DatabaseDog(1, "Klaus")
    whatDog(dog)  // prints "database dog"
}
```

In this example, changing the type annotation of the dog variable will change what is printed:
fun test() {
    val dog: Dog = DatabaseDog(1, "Klaus")
    whatDog(dog)  // prints "dog"
}

In my opinion this is very similar to the "issue" you described, and I don't see a big problem with it.

Reply:
Ah, yes, I forgot about function overloading. I would personally find the kind of overload exemplified by `whatDog()` to be a code smell for the same reasons as for why I am against overlapping instances, but I can see how it would not be as inconsistent with the current behavior of the language as I first thought it was. In my opinion, ambiguous overloads should also be disallowed, or at least result in a warning, but I acknowledge that the discussion of that is out of scope for this KEEP.

I can see how you don't have the transitivity issue when equals is statically resolved as an overload because then you technically implement two different equivalence relations without trying to combine them into one (which is impossible). I still find it confusing though, and something to be avoided. For example:
```
fun eq(dog1: Dog, dog2: Dog) = dog1.name == dog2.name
fun eq(dog1: DatabaseDog, dog2: DatabaseDog) = dog1.name == dog2.name && dog1.id == dog2.id
fun main() {
    val dog1 = DatabaseDog("Klaus", 1)
    val dog2 = ApiDog("Klaus")
    val dog3 = DatabaseDog("Klaus", 2)
    println(eq(dog1, dog2)) //true
    println(eq(dog2, dog3)) //true
    println(eq(dog1, dog3)) //false -- could just as well be true if the compiler resolved the ambiguity otherwise
}
```
Yes, it fixes the problem because both `eq` functions implement proper equivalence relations. But I think the code would be much clearer if the functions were simply named differently and if it didn't rely on the compiler resolving the ambiguity.

Reply:
@ulrikrasmussen
I think you raised some interesting points! 

@fatjoem
I am convinced that you pointed out the key: The behavior of the `eq` function is decided at compile time for the specified expression. This differentiates it from the "common mistakes" made with the `equals` function and removes (almost) all ambiguity.

I think the given example here

 ```kotlin
fun eq(dog1: Dog, dog2: Dog) = dog1.name == dog2.name
fun eq(dog1: DatabaseDog, dog2: DatabaseDog) = dog1.name == dog2.name && dog1.id == dog2.id
fun main() {
    val dog1 = DatabaseDog("Klaus", 1)
    val dog2 = ApiDog("Klaus")
    val dog3 = DatabaseDog("Klaus", 2)
    println(eq(dog1, dog2)) //true
    println(eq(dog2, dog3)) //true
    println(eq(dog1, dog3)) //false -- could just as well be true if the compiler resolved the ambiguity otherwise
}
```

is pretty interesting, since it has the same traits the proposed `Equality` type class by being resolved entirely at compile time. Still, it is _somewhat_ ambiguous since it seems like transitivity is broken at the call site. But honestly, I cannot see how one would be able to forbid something like this in the current type system (and with compile-time extension interfaces) and I am convinced that it is still much better than the current `Any.equals`, since it is possible to resolve problems arising with this function at the call site of the function and not deep inside of some implementation of some model in some type hierarchy. It's easy to here to specify the type class to use explicitly, change the type of dog (like proposed above) or specify the functions type variable!

```kotlin

fun solution1() {
    val databaseDog: DatabaseDog = getSomeDatabaseDog()
    val databaseDog2: DatabaseDog = getSomeDatabaseDog()

    databaseDog.eq<Dog>(databaseDog2)
}

fun solution2() {
    val databaseDog: Dog = object : DatabaseDog {}
    val databaseDog2: Dog = object : DatabaseDog {}

    databaseDog.eq(databaseDog2)
}

fun solution3() {
    val databaseDog: DatabaseDog = getSomeDatabaseDog()
    val databaseDog2: DatabaseDog = getSomeDatabaseDog()

    databaseDog.eq<Dog>(databaseDog2, Dog.Companion.DogEquality)
}
```




After all, I wished I would have been more creative when naming those example entities than `Dog`, `ApiDog` and `DatabaseDog` when using them for such discussions =)

Reply:
The only problem with those example entities is that it is hard to follow what we are talking about without having read the blog post.

Reply:
> But honestly, I cannot see how one would be able to forbid something like this in the current type system (and with compile-time extension interfaces)

Wouldn't that just amount to disallowing ambiguity at the call site?

Ambiguous method overloading serves no useful purpose while increasing the complexity of the code for no good reason. Going back to the example, in the third line
```
println(eq(dog1, dog3))
```
what is the programmer *actually* saying here? There are two possible overloads that match this one, namely `eq(Dog, Dog)` and `eq(DatabaseDog, DatabaseDog)`. The compiler picks the most specific one. So, this line actually tells the compiler "generate code that compares `dog1` and `dog3` by database identifier, *unless the implementation disappears, in which case compare them only by name*". I would argue that the "unless" part is **almost never** what the programmer intended, unless `eq(DatabaseDog, DatabaseDog)` is an optimized version of `eq(Dog, Dog)` with the same semantics. If `eq(DatabaseDog, DatabaseDog)` disappears in a refactoring, then the program should fail to compile rather than just arbitrarily change its semantics. The programmer has gained nothing but saved a few keystrokes - the program would be more robust if he had just given the two overloads distinct names (e.g. `eqDog` and `eqDatabaseDog`).

Overloaded (i.e. overlapping) extension classes are even worse, for two reasons:

1. Unlike method overloads, the overloads for an extension class are spread over multiple files instead of being defined in the same place. This makes it much harder for the programmer to get an overview of what is shadowing what.
2. Unlike method overloads, it is harder to provide good tooling which tells you what overload is in effect at a given place in a program. Yes, for extension classes without constraints, you can use "go to definition" to go the extension class that matches in a given context. But, suppose we had another extension interface (admittedly contrived, but not unrealistic):
```
extension interface EqList<A> { fun eqList(x: List<A>, y: List<A>): Bool }
extension class EqListImpl<A>(with val constraint: Equality<A>) { ... }
```
If I compare two lists of type `List<DatabaseDog>` with `eqList`, then there will only be one unique extension class, `EqListImpl` that matches. "Go to definition" in my IDE will tell me this by placing the cursor on the `EqListImpl` class. However, this class *has a constraint itself*, and this constraint *is* ambiguous because it can be matched by both `Equality<Dog>` and `Equality<DatabaseDog>`. But how do you actually communicate that in the IDE? When you used "go to definition" at the call site, the IDE placed the cursor at `EqListImpl`, so the scope of the actual call site has been left. To show what is actually in effect, the IDE would have to generate the concrete instantiation of the extension class chain, e.g. `EqListImpl(DatabaseDog.Companion.DogEquality)` and show you that. The compiler might even be disambiguating several layers of overlapping extension classes in a long chain.